crates/wysiwyg/src/composer_list_handler.rs

```diff
@@ -0,0 +1,190 @@
+        self.create_list(false)
+    }
+
+    pub(crate) fn do_enter_in_list(
+        &mut self,
+        parent_handle: DomHandle,
+        location: usize,
+        range: SameNodeRange,
+    ) -> ComposerUpdate<S> {
+        // Store current Dom
+        self.push_state_to_history();
+        let parent_node = self.state.dom.lookup_node(parent_handle.clone());
+        let list_node_handle = parent_node.handle().parent_handle();
+        if let DomNode::Container(parent) = parent_node {
+            if parent.is_empty_list_item() {
+                self.remove_list_item(
+                    list_node_handle,
+                    location,
+                    parent_handle.index_in_parent(),
+                    true,
+                );
+            } else {
+                self.slice_list_item(list_node_handle, location, range);
+            }
+            self.create_update_replace_all()
+        } else {
+            panic!("No list item found")
+        }
+    }
+
+    fn create_list(&mut self, ordered: bool) -> ComposerUpdate<S> {
+        // Store current Dom
+        self.push_state_to_history();
+        let list_tag = if ordered { "ol" } else { "ul" };
+        let (s, e) = self.safe_selection();
+        let range = self.state.dom.find_range(s, e);
+        match range {
+            Range::SameNode(range) => {
+                let node =
+                    self.state.dom.lookup_node(range.node_handle.clone());
+                if let DomNode::Text(t) = node {
+                    let text = t.data();
+                    let list_node = DomNode::new_list(
+                        S::from_str(list_tag),
+                        vec![DomNode::Container(ContainerNode::new_list_item(
+                            S::from_str("li"),
+                            vec![DomNode::Text(TextNode::from(text.clone()))],
+                        ))],
+                    );
+                    self.state.dom.replace(range.node_handle, vec![list_node]);
+                    return self.create_update_replace_all();
+                } else {
+                    panic!("Can't create a list from a non-text node")
+                }
+            }
+
+            Range::NoNode => {
+                self.state.dom.append_child(DomNode::new_list(
+                    S::from_str(list_tag),
+                    vec![DomNode::Container(ContainerNode::new_list_item(
+                        S::from_str("li"),
+                        vec![DomNode::Text(TextNode::from(S::from_str("")))],
+                    ))],
+                ));
+                return self.create_update_replace_all();
+            }
+
+            _ => {
+                panic!("Can't create ordered list in complex object models yet")
+            }
+        }
+    }
+
+    fn slice_list_item(
+        &mut self,
+        handle: DomHandle,
+        location: usize,
+        range: SameNodeRange,
+    ) {
+        assert_eq!(range.start_offset, range.end_offset);
```

Comment:
Why should these be equal? Maybe a comment explaining this function would help here.

Reply:
(a doc comment I mean)

Reply:
Actually it's in the wrong location, it should be in `do_enter_in_list`. I moved it there with a comment. (it's equal because we're expecting `enter` to delete current selection first then `do_enter_in_list` is called)